add a few batch rpc calls

Greg-Griffith opened the pull request:
Batch RPC calls added by this PR:
- getrawtransactions : get raw transaction data for every tx in a block
- getrawtransactionssince : getrawtransactions for a block and specified amount of its ancestors
- gettxoutproofs : provides a proof for each txid passed in instead of one proof for the entire set

Commit:
add getrawtransactions and getrawtransactionssince rpc calls

Commit:
fix getrawtransactionssince to properly get ancestors, if at least one block's information is found do not throw rpc error, simply return what was found. useful for pruned nodes if the rpc call puts them into pruned territory due to getting too many ancestors

Greg-Griffith commented:
@monsterbitar This is the PR i will be working with for the requested updates to batch some rpc calls. Two of the rpc calls we talked about last friday have been committed. gettxoutproofs is not needed because gettxoutproof can already handle multiple txids. I have not yet tested either of these new calls.

Greg-Griffith commented:
This PR will address some of the concerns in #1563

sickpig commented:
```
../../src/rpc/rawtransaction.cpp:444:21: error: use of undeclared identifier 'acestorcount'; did you mean 'ancestorcount'?
    int64_t limit = acestorcount + 1;
                    ^~~~~~~~~~~~
                    ancestorcount
../../src/rpc/rawtransaction.cpp:430:13: note: 'ancestorcount' declared here
    int64_t ancestorcount = 0;
```

monsterbitar commented:
If I have not misunderstood, gttxoutproof takes multiple transactions as indata but creates a single proof that all those transactions are included in a given block - this creates a larger proof than I need, as I need to be able to serve proofs for each TX separately.

monsterbitar commented:
getrawtransactions might be better named getrawblocktransactions as it is based on a block.

Greg-Griffith commented:
my mistake. i misunderstood what you needed from that call. Will add it today. will also rename getrawtransactions to getrawblocktransactions

Commit:
fix typo, getrawtransactions is now getrawblocktransactions

Commit:
implement gettxoutproofs

Commit:
formatting fixes

monsterbitar commented:
alright, looks like a good time for me to test this. I'm running a source distribution (gentoo) so I have all the dependencies needed to compile it. It says "all checks have failed" here though, should I wait a bit more so you can sort that out?

monsterbitar commented:
I have been doing performance measuring in my application and loading and parsing all blocks and transactions since the cashaccount activation takes ~4 minutes.

With ~3,800 blocks and ~257,000 transactions I'm doing ~270,000 RPC calls. The time needed to call getRawTransaction alone represents somewhere between 75% and 90% of my application performance cost and with getRawBlockTransactions I should be able to reduce the calls to ~4,000 instead of the current ~270,000.

Aught to remove a bit of RPC overhead, but I don't know how much time will be spent generating and delivering the arrays with rawTransactions so until this is ready and I manage to compile and test it I can't know what impact it will have.

sickpig commented:
@monsterbitar @Greg-Griffith 

since we got #1577 merged yesterday this patch has to be updated to use the new pushKV method.

Greg-Griffith commented:
Will rebase and update in just a second

Commit:
change push_back(Pair()) to pushKV

monsterbitar commented:
I cloned and compiled your fork @Greg-Griffith, and have the bitcoind running. The output of bitcoin-cli help is inconsistent with the implementation. The RPC call to getRawBlockTransactions however, seems to work just fine.

The end result is that I cut down from 270k RPC calls for a sync, to 12.5k - the time went from ~240s to ~100s. The vast majority of the work done in the cash account lookup server is still waiting for the blocks transactions to be delivered.

monsterbitar commented:
When I call getTxOutProofs with an array containing string with transaction hexes, and as the second parameter a single string with the blockhash hex I get an empty array returned.

Commit:
improve gettxoutproofs logic

monsterbitar commented:
With getTxOutProofs working, it takes roughtly the same time but with fewer RPC calls (from 12.5k to 11.6k calls).

Commit:
remove cs_main lock from getrawblocktransactions and getrawtransactionssince

Afaict we dont need to lock cs_main here. LookupBlockIndex has its own 
lock that it calls internally and ReadBlockFromDisk doesnt touch any in 
memory data structures.

 I need soomeone to verify if my conclusion on this is correct.

Commit:
rework getrawtransactionssince to move towards tip instead of genesis

fixed json structures returned by help for getrawblocktransactions and 
getrawtransactionssince

Commit:
formatting fix

Commit:
fix issue where pushKV requies VOBJ

Commit:
add gettxoutproofs to merkleblock.py test

Commit:
add python test for getrawtransactionssince and getrawblocktransactions

Greg-Griffith commented:
The content of this PR is done. Requesting additional reviewers

gandrewstone commented on src/rpc/rawtransaction.cpp:
```diff
@@ -218,6 +218,252 @@ UniValue getrawtransaction(const UniValue &params, bool fHelp)
+{
+    if (fHelp || params.size() < 1 || params.size() > 2)
+        throw runtime_error(
+            "getrawtransactions \"blockhash\" ( verbose )\n"
```
help needs to be "getrawblocktransactions".

I would like to move away from the boolean flag system that some bitcoin RPCs currently use because it is not user friendly.  Although it may be a little weird when writing code, I propose we use the command line convention of "-flag" or "--flag=value" (all the dash options MUST be optional), placed before the other args.  Another possibility is to use "flag=value" which is kind of a compromise between the programmatic way of specifying keyword arguments and the command line.

Greg-Griffith replied:
will fix that error in help.

gandrewstone commented on src/rpc/rawtransaction.cpp:
```diff
@@ -218,6 +218,252 @@ UniValue getrawtransaction(const UniValue &params, bool fHelp)
+        }
+
+        UniValue result(UniValue::VOBJ);
+        result.pushKV("hex", strHex);
```
What about not providing the hex if verbose is set?  Why would someone want hex and verbose, and is this common enough to optimize it as opposed to having the user call the API 2x? 

If we drop the hex, this will speed things up because the message will be smaller and there will be less JSON encode/decode.

Greg-Griffith replied:
I left the hex in there to be consistent with getrawtransaction which also returns the hex regardless of verbosity. i have no problem removing it from all of the calls but i feel some feedback from app devs should be collected first.

gandrewstone commented on src/rpc/rawtransaction.cpp:
```diff
@@ -218,6 +218,252 @@ UniValue getrawtransaction(const UniValue &params, bool fHelp)
+    if (fHelp || params.size() < 1 || params.size() > 3)
+        throw runtime_error(
+            "getrawtransactionssince \"blockhash\" ( verbose ) ( count )\n"
+            "\nReturn the raw transaction data for a <count> blocks starting with blockhash and moving towards the "
```
remove extra "a"

Is "count" blocks what the user wants, or does he really want approximately N bytes of data?  If this API is meant to chunk the data, it would make more sense to specify N bytes and we give him M complete blocks such that Len(message(M)) >= N.

Greg-Griffith replied:
This was implemented as requested. This was a way to batch multiple getrawblocktransactions into one call. 
Overall the process moved from getrawtransaction for every tx in a block every block. to getrawblocktransactions for one block or getrawtransactionssince for multiple blocks. 

although i see now how the naming convention is probably not the best to describe the use case

gandrewstone commented on src/rpc/rawtransaction.cpp:
```diff
@@ -218,6 +218,252 @@ UniValue getrawtransaction(const UniValue &params, bool fHelp)
+            HelpExampleCli("getrawtransactionssince", "\"hashblock\" 1 10") +
+            HelpExampleRpc("getrawtransactionssince", "\"hashblock\", 1 10"));
+
+    LOCK(cs_main);
```
you are right that cs_main needs to be held here in case of a reorg, so that we return a set of blocks that is consistent.  However, note that a reorg could happen moments after this call -- the caller still needs to handle that case.

So one way to minimize cs_main time is only lock it while copying block hashes from chainActive to a temporary array.

Then you can unlock during the lengthy period where you read each block off of the disk and process it.

optionally, at the very end check that chainActive[lastblock] == last block in your array, and if it does not throw out your results (or part of them) and recalculate.

gandrewstone commented on src/rpc/rawtransaction.cpp:
```diff
@@ -302,6 +548,86 @@ UniValue gettxoutproof(const UniValue &params, bool fHelp)
+{
+    if (fHelp || params.size() != 2)
+        throw runtime_error(
+            "gettxoutproofs [\"txid\",...] ( blockhash )\n"
```
if blockhash is optional, use [], since txid is not optional, use ()

EDIT WRT txid, sorry I was not realizing that [txid...] was meant to be a JSON list.  I am not immediately aware of other RPC apis that take lists, how do you call this on the command line?

Greg-Griffith replied:
the reason this takes a JSON list is because arg is a direct copy of gettxoutproof's args which also takes a JSON list. block hash is no longer optional though

gandrewstone commented:
@monsterbitar how do you use gittxoutproof?  Is it always for new, recent tx, for random tx, for prevouts?  Are you getting a lot of tx that you know are in 1 block, or are you getting a lot of tx from a lot of blocks?

BTW, WRT CoinAccessor, its a shared lock that's being taken so it will only block if the utxo is currently being updated.

Commit:
fix incorrect help message for getrawblocktransactions

Commit:
remove extra "a" in getrawtransactionssince help message

monsterbitar commented:
@gandrewstone "how do you use gittxoutproof?"

I use getrawblocktransactions for a given block, iterate the transactions and discard all that are not cashaccount registrations (in ~500k tx, so far, only 8k is registrations), then request the proofs for all registrations for that block before moving on to the next block.

Commit:
use verbose flag (--verbose or -v) instead of num to represent bool

Commit:
update getrawtransaction.py test for new verbose usage. fix formatting

Commit:
add protocol id specific search

Commit:
make HasData a const

Commit:
add hasdatas check to getrawblocktransactions and getrawtransactionssince tests

Greg-Griffith commented:
@monsterbitar i added a new feature that lets you search for a specific protocol ID inside a block by submitting an added param at the end. Would you find something like this useful?

monsterbitar commented:
I would, it moves some of my logic over to your end. how is the syntax - how do I use it?

Greg-Griffith commented:
@monsterbitar the params for the functions changed in that last PR. example py script is getrawtransaction.py

getrawblocktransactions is now:
            "1. \"-v\" or \"--verbose\" 
            "2. \"hashblock\" 
            "3. \"protocol_id\" 

so to use verbose you simply add a -v at the beginning similar to what you would do for cmdline
example: getrawblocktransactions("-v", "blockhash")
the 3rd param would be a protocol id, "*" is a wild card for any id
example: getrawblocktransactions("blockhash" , "*")
or you can specify
example: getrawblocktransactions("blockhash" , "1234")

getrawtransactionssince has similar changes where protocolid was added as the last arg after count.
            "1. \"-v\" or \"--verbose\" 
            "2. \"hashblock\" 
            "3. count (numeric)" 
            "4. \"protocol_id\" 
 examples being:
getrawtransactionssince("-v", "blockhash" , 1, "*")
getrawtransactionssince("blockhash" , 1, "*")
getrawtransactionssince("blockhash" , 1, "1234")

gandrewstone commented:
@monsterbitar, I'm going to merge this now, but if you have any additional API ideas let us know.  Until we make a release, we can still tweak the RPC API without worrying about backwards compatibility.

stiffpoo commented:
having issues defining the # of blocks since in bitcoin-cli. taking it as a string input, not an integer as desired (using v1.6 ubuntu x64)
```
bitcoin-cli getrawtransactionssince "blockhash" 10
error code: -1
error message:
JSON value is not an integer as expected
```

Greg-Griffith commented:
@stiffpoo thanks for bringing this to my attention. it should be fixed in #1711 

Next time please open a new issue instead of commenting on a closed PR. It will help us to see the issue faster. Thank you

stiffpoo commented:
No worries, good to hear.

Yeah apologies for the run around, my bad.
